### ISSUE TITLE
Provide the destination object in a PreCondition overload

### ISSUE THREAD

**Issue by jez9999** [AutoMapper/AutoMapper#2727]
I want to implement a side-effect during my `PreCondition` mapping of setting a boolean on the destination object, and it has to be done at the `PreCondition` phase because it depends on the order of the objects being traversed starting from the initial object and expanding into referenced objects, in that order.  At the moment there is no overload lambda of `PreCondition` that gives me access to the destination object.  Could we add one?

**Comment by lbargaoanu** [comment]
A PR is welcome. Similar to [this](https://github.com/AutoMapper/AutoMapper/blob/8a4e0bcc3f0cbb2337fd5b99e44432b2f415cfae/src/AutoMapper/Execution/TypeMapPlanBuilder.cs#L477).

**Comment by jez9999** [comment]
Yeah, I was thinking it looked pretty simply to pass in `_destination` - so much so that I wondered whether there was an architectural reason why that overload didn't exist.  Like, is it just building a graph of booleans of "whether to map this member" at the `PreCondition` stage, so the destination object hasn't been created yet, or something?  Or is it just an oversight?

**Comment by lbargaoanu** [comment]
Check [the execution plan](http://docs.automapper.org/en/stable/Understanding-your-mapping.html). But other than that, it seems fair that who needs the overload should implement it :)

**Comment by jez9999** [comment]
Created PR.
